finality-aleph/src/sync/message_limiter.rs

```diff
@@ -0,0 +1,188 @@
+use std::fmt::{Display, Formatter};
+
+use parity_scale_codec::Encode;
+
+use crate::sync::data::MAX_SYNC_MESSAGE_SIZE;
+
+const MSG_BYTES_LIMIT: usize = MAX_SYNC_MESSAGE_SIZE as usize;
+
+pub struct Limiter<'a, D: Encode, const LIMIT: usize> {
+    msg: &'a [D],
+    start_index: usize,
+    indexes: Vec<usize>,
+}
+
+pub type MsgLimiter<'a, D> = Limiter<'a, D, MSG_BYTES_LIMIT>;
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    ItemTooBig,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::ItemTooBig => write!(f, "Single item takes more than the limit"),
+        }
+    }
+}
+
+impl<'a, D: Encode, const LIMIT: usize> Limiter<'a, D, LIMIT> {
+    pub fn new(msg: &'a [D]) -> Self {
+        Self {
+            msg,
+            start_index: 0,
+            indexes: (0..=msg.len()).collect(),
+        }
+    }
+
+    pub fn next_largest_msg(&mut self) -> Result<Option<&'a [D]>, Error> {
+        if self.start_index == self.msg.len() {
+            return Ok(None);
+        }
+        let end_idx = self.bs_by_encode()?;
+
+        let start_index = self.start_index;
+        self.start_index = end_idx;
+
+        Ok(Some(&self.msg[start_index..end_idx]))
+    }
+
+    fn bs_by_encode(&self) -> Result<usize, Error> {
+        let indexes = &self.indexes[self.start_index..];
+
+        let idx = indexes.partition_point(|&idx| {
+            let encoded_size = self.msg[self.start_index..idx].encoded_size();
```

Comment:
Does it `panic` when called one more time after returning last element? Maybe we can make it keep returning `None` or move back to previous `Iterator` implementation.

Reply:
it keeps returning None or Err(e)